drivers/usb/uhc/uhc_mcux_common.c

```diff
@@ -0,0 +1,252 @@
+
+#include "uhc_common.h"
+#include "usb.h"
+#include "usb_host_config.h"
+#include "usb_host_mcux_drv_port.h"
+#include "uhc_mcux_common.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(uhc_mcux, CONFIG_UHC_DRIVER_LOG_LEVEL);
+
+#define PRV_DATA_HANDLE(_handle) CONTAINER_OF(_handle, struct uhc_mcux_data, mcux_host)
+
+int uhc_mcux_lock(const struct device *dev)
+{
+	struct uhc_data *data = dev->data;
+
+	return k_mutex_lock(&data->mutex, K_FOREVER);
+}
+
+int uhc_mcux_unlock(const struct device *dev)
+{
+	struct uhc_data *data = dev->data;
+
+	return k_mutex_unlock(&data->mutex);
+}
+
+int uhc_mcux_control(const struct device *dev, uint32_t control, void *param)
+{
+	struct uhc_mcux_data *priv = uhc_get_private(dev);
+	usb_status_t status;
+
+	status = priv->mcux_if->controllerIoctl(priv->mcux_host.controllerHandle, control, param);
+	if (status != kStatus_USB_Success) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+int uhc_mcux_bus_control(const struct device *dev, usb_host_bus_control_t type)
+{
+	return uhc_mcux_control(dev, kUSB_HostBusControl, &type);
+}
+
+/* Controller driver calls this function when device attach. */
+usb_status_t USB_HostAttachDevice(usb_host_handle hostHandle, uint8_t speed, uint8_t hubNumber,
+				  uint8_t portNumber, uint8_t level,
+				  usb_device_handle *deviceHandle)
+{
+	enum uhc_event_type type;
+	struct uhc_mcux_data *priv;
+
+	if (speed == USB_SPEED_HIGH) {
+		type = UHC_EVT_DEV_CONNECTED_HS;
+	} else if (speed == USB_SPEED_FULL) {
+		type = UHC_EVT_DEV_CONNECTED_FS;
+	} else {
+		type = UHC_EVT_DEV_CONNECTED_LS;
+	}
+
+	priv = (struct uhc_mcux_data *)(PRV_DATA_HANDLE(hostHandle));
+	uhc_submit_event(priv->dev, type, 0);
+
+	return kStatus_USB_Success;
+}
+
+/* Controller driver calls this function when device detaches. */
+usb_status_t USB_HostDetachDevice(usb_host_handle hostHandle, uint8_t hubNumber, uint8_t portNumber)
+{
+	struct uhc_mcux_data *priv;
+
+	priv = (struct uhc_mcux_data *)(PRV_DATA_HANDLE(hostHandle));
+	uhc_submit_event(priv->dev, UHC_EVT_DEV_REMOVED, 0);
+	uhc_mcux_bus_control(priv->dev, kUSB_HostBusEnableAttach);
+
+	return kStatus_USB_Success;
+}
+
+/* MCUX Controller driver calls this function to get the device information. */
+usb_status_t USB_HostHelperGetPeripheralInformation(usb_device_handle deviceHandle,
+						    uint32_t infoCode, uint32_t *infoValue)
+{
+	/* The deviceHandle is struct uhc_transfer because Zephyr environment doesn't allow
+	 * to call upper layer API to get usb device information from controller driver.
+	 * The device information should be able to obtain from the struct uhc_transfer.
+	 */
+	struct uhc_transfer *const xfer = (struct uhc_transfer *const)deviceHandle;
+	usb_host_dev_info_t info_code;
+
+	if ((deviceHandle == NULL) || (infoValue == NULL)) {
+		return kStatus_USB_InvalidParameter;
+	}
+	info_code = (usb_host_dev_info_t)infoCode;
+	switch (info_code) {
+	case kUSB_HostGetDeviceAddress:
+		*infoValue = (uint8_t)xfer->addr;
```

Comment:
addr has benn removed in favor of `struct usb_device *udev`
`		*infoValue = xfer->udev->addr;`

Reply:
updated.